### PR TITLE
Implement OpenID Provider Configuration endpoint

### PR DESCRIPTION
**Pull request by Kehrlann**
- See https://openid.net/specs/openid-connect-discovery-1_0.html
  sections 3 and 4.
- We introduce here a "ProviderSettings" construct to configure
  the authorization server, starting with endpoint paths (e.g.
  token endpoint, jwk set endpont, ...)

This will deliver #55

### CODE REVIEW

**Comment by Kehrlann** [comment]
I have introduced `org.springframework.security.oauth2.core.ClientAuthenticationMethod2` so that the names match "xxx_auth_methods_supported` in OpenID Discovery 1.0. 

As a side note, It does not derive from `org.springframework.security.oauth2.core.ClientAuthenticationMethod` in spring-security because that class is final...

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see ProviderSettings
+ * @see <a target="_blank" href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig">OpenID Connect Discovery 1.0</a>
+ */
+public class OidcProviderConfigurationFilter extends OncePerRequestFilter {
+	private final RequestMatcher requestMatcher;
+	private final OidcProviderConfiguration providerConfiguration;
+	private final OidcProviderConfigurationHttpMessageConverter messageConverter = new OidcProviderConfigurationHttpMessageConverter();
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java:54]: Rename to `providerConfigurationHttpMessageConverter`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+			throws ServletException, IOException {
+		if (!this.requestMatcher.matches(request)) {
+			filterChain.doFilter(request, response);
+			return;
+		}
+
+		ServletServerHttpResponse resp = new ServletServerHttpResponse(response);
+		messageConverter.write(this.providerConfiguration, MediaType.APPLICATION_JSON, resp);
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java:85]: Prefix `this`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * A facility for OpenID Connect Provider Configuration settings.
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see Settings
+ * @see <a target="_blank" href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata">OpenID Connect Discovery 1.0</a>
+ */
+public class ProviderSettings extends Settings {
+	private static final String PROVIDER_SETTING_BASE = "spring.security.oauth2.authorization-server.provider.";
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java:35]: Change prefix value to `setting.provider.`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	public static final String ISSUER = PROVIDER_SETTING_BASE.concat("issuer");
+	public static final String OPENID_CONFIGURATION_ENDPOINT = PROVIDER_SETTING_BASE.concat("openid-configuration-endpoint");
+	public static final String AUTHORIZATION_ENDPOINT = PROVIDER_SETTING_BASE.concat("authorization-endpoint");
+	public static final String TOKEN_ENDPOINT = PROVIDER_SETTING_BASE.concat("token-endpoint");
+	public static final String JWK_SET_ENDPOINT = PROVIDER_SETTING_BASE.concat("jwk-set-endpoint");
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java:40]: Change value to `jwks-endpoint`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	public URL issuer() {
+		return setting(ISSUER);
+	}
+
+	/**
+	 * Returns the OpenID Provider Configuration endpoint. The default is {@code /.well-known/openid-configuration}.
+	 *
+	 * @return the OpenID Provider Configuration endpoint
+	 */
+	public String openIDProviderConfigurationEndpoint() {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java:74]: Let's remove this as the spec states it should be `/.well-known/openid-configuration`. I'm not sure if we need to allow this endpoint to be customized.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the PR @Kehrlann. Please see review comments.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	private final OidcProviderConfiguration providerConfiguration;
+	private final OidcProviderConfigurationHttpMessageConverter messageConverter = new OidcProviderConfigurationHttpMessageConverter();
+
+	public OidcProviderConfigurationFilter(ProviderSettings providerSettings) {
+		Assert.notNull(providerSettings, "providerSettings cannot be null");
+		this.requestMatcher = new AntPathRequestMatcher(
+				providerSettings.openIDProviderConfigurationEndpoint(),
+				HttpMethod.GET.name()
+		);
+		this.providerConfiguration = new OidcProviderConfiguration.Builder()
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java:62]: We should build `OidcProviderConfiguration` per request given that `ProviderSettings` may change at runtime since it's mutable.

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java [hunk 7]
```diff
@@ -249,4 +274,17 @@ public void configure(B builder) {
 	private static <B extends HttpSecurityBuilder<B>> KeyManager getKeyManagerBean(B builder) {
 		return builder.getSharedObject(ApplicationContext.class).getBean(KeyManager.class);
 	}
+
+	private static <B extends HttpSecurityBuilder<B>> ProviderSettings getProviderSettings(B builder) {
+		ProviderSettings providerSettings = builder.getSharedObject(ProviderSettings.class);
+		if (providerSettings == null) {
+			providerSettings = getProviderSettingsBean(builder);
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:281]: `ProviderSettings` will not be a required `@Bean` since we can default the values - `getProviderSettingsBean()` will throw if it's not defined

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java [hunk 8]
```diff
@@ -80,6 +82,8 @@
 			OAuth2TokenRevocationEndpointFilter.DEFAULT_TOKEN_REVOCATION_ENDPOINT_URI, HttpMethod.POST.name());
 	private final RequestMatcher jwkSetEndpointMatcher = new AntPathRequestMatcher(
 			JwkSetEndpointFilter.DEFAULT_JWK_SET_ENDPOINT_URI, HttpMethod.GET.name());
+	private final RequestMatcher oidcProviderMetadataEndpointMatcher = new AntPathRequestMatcher(
+			"/.well-known/openid-configuration", HttpMethod.GET.name());
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:86]: Let's add "/.well-known/openid-configuration" as a constant in the `Filter`, same as all other `Filter`'s

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcProviderMetadataClaimAccessor.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+	/**
+	 * Returns the Subject Identifier types supported {@code (subject_types_supported)}.
+	 *
+	 * @return the Subject Identifier types supported
+	 */
+	default List<String> getSubjectTypes() {
+		return this.getClaimAsStringList(OidcProviderMetadataClaimNames.SUBJECT_TYPES_SUPPORTED);
+	}
+
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcProviderMetadataClaimAccessor.java:109]: Good point. ID_TOKEN_SIGNING_ALG_VALUES_SUPPORTED is not used yet. Should we remove it from the claim names instead or provide access ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcProviderConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see OidcProviderMetadataClaimAccessor
+ * @see <a target="_blank" href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata">OpenID Connect Discovery 1.0</a>
+ */
+public class OidcProviderConfiguration implements OidcProviderMetadataClaimAccessor {
+	private final Map<String, Object> claims;
+
+	private OidcProviderConfiguration(Map<String, Object> claims) {
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcProviderConfiguration.java:44]: Should we somehow validate the claims ? The initial pattern was to make it private and force the use of the builder, which validates the claims.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java (hunk 6 above)
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OidcProviderConfigurationFilter.java:62]: If the Issuer URL is invalid (e.g. with spaces inside), what's the type of the exception we want to throw ?

It's an edge case, we'll be wrapping `MalformedURLException`.

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java (hunk 8 above)
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:86]: Done.

Most of the other filters endpoints will be configurable through ProviderSettings though. The defaults are also in the ProviderSettings (hardcoded as well). Is there something specific we want to do about this ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcProviderConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+		}
+
+		/**
+		 * Constructs a new {@link Builder} from the provided claims.
+		 *
+		 * @param claims the claims
+		 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata">OpenID Connect Discovery 1.0</a>
+		 * for required claims
+		 */
+		public Builder(Map<String, Object> claims) {
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcProviderConfiguration.java]: See comment about validating the claims in `OidcProviderConfiguration`. For now, only the Builder validates the claims, and this builder constructor is used when reading a response through `OidcProviderConfigurationHttpMessageConverter`.

File: oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java (hunk 7 above)
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerConfigurer.java:281]: I like the idea.

If we don't register the filter at all, and the user navigates to `/.well-known/openid-configuration` then they get an `HTTP 302` redirecting to `/login`. I'm feeling we should instead return an `HTTP 404`. Thoughts ?

Additionally, `ProviderSettings` is mutable and could change. If the issuer is added after startup, we wouldn't be registering the endpoint. Is that fine ?

I'd love a bit of guidance on writing some tests for that.

**Comment by Kehrlann** [comment]
I've added integration tests to `OAuth2AuthorizationCodeGrantTests`, but I'm not sure it's the best place for it. Maybe it should go in it's own `OidcProviderConfiugrationTests`, or maybe a broader `OidcTests` for oidc-related workflows ?

**Comment by joshuatcasey** [comment]
Hi from the Application SSO team. We had a team goal to implement a customizable Issuer URI - thanks for beating us to it!

We noticed a few things with `64207e9`.

https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata

> issuer
REQUIRED. URL using the https scheme with no query or fragment component that the OP asserts as its Issuer Identifier. If Issuer discovery is supported (see Section 2), this value MUST be identical to the issuer value returned by WebFinger. This also MUST be identical to the iss Claim value in ID Tokens issued from this Issuer.

https://tools.ietf.org/html/rfc8414#section-2

**Comment by Kehrlann** [comment]
Hi @joshuatcasey ! Glad to help. To your points, here are some of the discussions we had around these topics:
1. **`issuer` not required:** The main idea for keeping the issuer optional (for now) is to stand up an authz server with absolute minimal configuration, even if it doesn't implement all possible extensions. I'll let @jgrandja chime in on this.
1. **using the `issuer` in tokens:** Yes the `iss` will be populated from the ProviderSettings, it is on the roadmap, not in the scope of this PR though.
1. **`issuer` validation:** That's correct, for local development `https` is a bit of a pain. We talked about pushing the validation of the URL to the users of the lib instead.

**Comment by joshuatcasey** [comment]
Thanks for the clarifications.

1. Understood. It will likely be required for app SSO, but I understand if that's not part of the library.
1. Can you link this on the roadmap? I see that #148 was closed in favor of this PR. Happy to submit a PR for this soon.
1. Yeah, we had the same thought about local development or deployments where "TLS everywhere" isn't enforced. We'll probably add a flag to allow `http` as well as `https` and make it very obvious to the configurer that this is unsafe / outside of specification compliance.

**Comment by Kehrlann** [comment]
I think @alek-sys will be making sure the `iss` claim is correctly populated, probably based off of #145

**Comment by alek-sys** [comment]
> I think @alek-sys will be making sure the `iss` claim is correctly populated, probably based off of #145 (that PR brings some changes to the API as well)

#145 will abstract JWT token creation behind `TokenIssuer` interface, but current the default implementation for JWT access token doesn't have any validations, not even for `iss` claim. There used to be a mechanism to validate issued token, but I couldn't find any use of it so after chatting with Joe decided to delete it. Maybe `iss` token schema validation is a reason to bring it back.

**Comment by joshuatcasey** [comment]
Speaking as an application developer, I want to validate the user's configuration on application startup, so it makes sense that we would validate `issuer URL` at that time. For now, we'll have some custom logic to perform this validation and we can discuss pushing it upstream to this library later.

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2ClientCredentialsGrantTests.java [hunk 12]
```diff
@@ -70,10 +74,11 @@
 	private MockMvc mvc;
 
 	@BeforeClass
-	public static void init() {
+	public static void init() throws MalformedURLException {
 		registeredClientRepository = mock(RegisteredClientRepository.class);
 		authorizationService = mock(OAuth2AuthorizationService.class);
 		keySource = new StaticKeyGeneratingCryptoKeySource();
+		providerSettings = new ProviderSettings().issuer(new URL("https://example.com"));
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2ClientCredentialsGrantTests.java:81]: Please revert the changes as `providerSettings` is not needed in these tests

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2RefreshTokenGrantTests.java [hunk 13]
```diff
@@ -76,10 +80,11 @@
 	private MockMvc mvc;
 
 	@BeforeClass
-	public static void init() {
+	public static void init() throws MalformedURLException {
 		registeredClientRepository = mock(RegisteredClientRepository.class);
 		authorizationService = mock(OAuth2AuthorizationService.class);
 		keySource = new StaticKeyGeneratingCryptoKeySource();
+		providerSettings = new ProviderSettings().issuer(new URL("https://example.com"));
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2RefreshTokenGrantTests.java:87]: Please revert the changes as `providerSettings` is not needed in these tests

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2TokenRevocationTests.java [hunk 14]
```diff
@@ -75,10 +79,11 @@
 	private MockMvc mvc;
 
 	@BeforeClass
-	public static void init() {
+	public static void init() throws MalformedURLException {
 		registeredClientRepository = mock(RegisteredClientRepository.class);
 		authorizationService = mock(OAuth2AuthorizationService.class);
 		keySource = new StaticKeyGeneratingCryptoKeySource();
+		providerSettings = new ProviderSettings().issuer(new URL("https://example.com"));
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2TokenRevocationTests.java:86]: Please revert the changes as `providerSettings` is not needed in these tests

File: samples/boot/oauth2-integration/authorizationserver/src/main/java/sample/config/AuthorizationServerConfig.java [hunk 15]
```diff
@@ -64,6 +67,15 @@ public CryptoKeySource keySource() {
 		return new StaticKeyGeneratingCryptoKeySource();
 	}
 
+	@Bean
+	public ProviderSettings providerSettings() {
```
Comment (jgrandja) [inline samples/boot/oauth2-integration/authorizationserver/src/main/java/sample/config/AuthorizationServerConfig.java:71]: Please revert the changes as `ProviderSettings` is not required

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the updated @Kehrlann. Please see review comments.

**Comment by jgrandja** [comment]
@joshuatcasey 

> I want to validate the user's configuration on application startup

Agreed. We will validate the application supplied configuration at startup and fail-fast on errors. I'm not sure where this will go at the moment but it very well might  go in `OAuth2AuthorizationServerConfigurer`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcProviderConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+
+		private Builder() { }
+
+		/**
+		 * Use this {@code issuer} in the resulting {@link OidcProviderConfiguration}, REQUIRED.
+		 *
+		 * @param issuer the issuer URI
+		 * @return the {@link Builder} for further configuration
+		 */
+		public Builder issuer(URL issuer) {
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcProviderConfiguration.java:100]: Good idea. Questions:
- We keep the claims as URLs, though, correct ?
- And therefore we should validate the URLs upon calling `.build()` ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+		return setting(ISSUER);
+	}
+
+	/**
+	 * Sets the URL the Provider uses as its Issuer Identity.
+	 *
+	 * @param issuer the URL the Provider uses as its Issuer Identity.
+	 * @return the {@link ProviderSettings} for further configuration
+	 */
+	public ProviderSettings issuer(URL issuer) {
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java:63]: Should we run the validation in this PR ?
If so, where exactly ? In `OAuth2AuthorizationServerConfigurer#init` ?

[the remark this answers is not in the document]
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/config/ProviderSettings.java:63]: Yes, let's validate in `OAuth2AuthorizationServerConfigurer#init`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcProviderConfiguration.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcProviderConfiguration.java:103]: > We keep the claims as URLs, though, correct

Yes, getters stay as-is. Just change the setters in builder.

> And therefore we should validate the URLs upon calling .build()

I don't think we need to validate here since we're validating `ProviderSettings` in `OAuth2AuthorizationServerConfigurer`

[the remark this answers is not in the document]
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/oidc/OidcProviderConfiguration.java:103]: We are validating they're not null and that they're an instance of URL. If we put Strings we could validate they can be cast to URLs.

File: samples/boot/oauth2-integration/authorizationserver/src/main/java/sample/config/AuthorizationServerConfig.java (hunk 15 above)
Comment (Kehrlann) [reply samples/boot/oauth2-integration/authorizationserver/src/main/java/sample/config/AuthorizationServerConfig.java:71]: Should we keep it anyway ? Otherwise we wouldn't have the OpenID Provider Configuration endpoint available (and later OAuth 2.0 Server Metadata endpoint).

We could add a comment explaining _why_ we have this Bean.

Comment (jgrandja) [reply samples/boot/oauth2-integration/authorizationserver/src/main/java/sample/config/AuthorizationServerConfig.java:71]: Let's remove it since we don't have OIDC auth implemented yet #53. I'll be starting #53 next week and I'll be sure to add this back to the sample as part of that PR.

**Comment by Kehrlann** [comment]
Integrated comments.

Could you please take a look at `OAuth2AuthorizationServerConfigurer#validateProviderSettings` for validating the issuer URI ?
Also added `ObjectToSetStringConverter2` here until we add it to Spring Security ; could you please take a look at it before integrating it ? I used `ObjectToListStringConverter` as an inspiration, but I only added temporary tests for `ObjectToSetStringConverter2#convert`, I am a bit unsure about `#getConvertibleTypes` and `#matches`.

**Review by jgrandja — APPROVED** [review]

**Comment by jgrandja** [comment]
Great work @Kehrlann ! This is now in master!
